dlib/serialize.h

```diff
@@ -1583,47 +1584,86 @@ namespace dlib
 
 // ----------------------------------------------------------------------------------------
 
-    class proxy_serialize 
+    class proxy_serialize
     {
     public:
         explicit proxy_serialize (
             const std::string& filename
-        ) 
+        ) : fout_optional_owning_ptr(new std::ofstream(filename.c_str(), std::ios::binary)),
+            fout(*fout_optional_owning_ptr)
         {
-            fout.reset(new std::ofstream(filename.c_str(), std::ios::binary));
-            if (!(*fout))
+            if (!fout)
                 throw serialization_error("Unable to open " + filename + " for writing.");
         }
-    
+        
+        explicit proxy_serialize (
+            std::ostringstream& ss
+        ) : fout_optional_owning_ptr(nullptr),
+            fout(ss)
+        {}
+        
+        explicit proxy_serialize (
+            std::stringstream& ss
+        ) : fout_optional_owning_ptr(nullptr),
+            fout(ss)
+        {}
+        
+        explicit proxy_serialize (
+            std::vector<char>& buf
+        ) : fout_optional_owning_ptr(new vectorstream(buf)),
+            fout(*fout_optional_owning_ptr)
+        {
+        }
+        
         template <typename T>
         inline proxy_serialize& operator<<(const T& item)
         {
-            serialize(item, *fout);
+            serialize(item, fout);
             return *this;
         }
 
     private:
-        std::shared_ptr<std::ofstream> fout;
+        std::unique_ptr<std::ostream> fout_optional_owning_ptr;
+        std::ostream& fout;
     };
-
-    class proxy_deserialize 
+    
+    class proxy_deserialize
     {
     public:
         explicit proxy_deserialize (
             const std::string& filename
-        )  : filename(filename)
+        )  : fin_optional_owning_ptr(new std::ifstream(filename.c_str(), std::ios::binary)),
+             fin(*fin_optional_owning_ptr)   
         {
-            fin.reset(new std::ifstream(filename.c_str(), std::ios::binary));
-            if (!(*fin))
+            if (!fin)
                 throw serialization_error("Unable to open " + filename + " for reading.");
-
-            // read the file header into a buffer and then seek back to the start of the
-            // file.
-            fin->read(file_header,4);
-            fin->clear();
-            fin->seekg(0);
+            init();
         }
-
+        
+        explicit proxy_deserialize (
+            std::istringstream& ss
+        ) : fin_optional_owning_ptr(nullptr),
+            fin(ss)
+        {
+            init();
+        }
+        
+        explicit proxy_deserialize (
+            std::stringstream& ss
```

Comment:
Same here.  This isn't needed if the constructor above just takes `std::istream`

Reply:
Yep